dashboard/app/models/lesson.rb

```diff
@@ -723,6 +723,91 @@ def lesson_plan_has_verified_resources
     resources.any? {|r| r.audience == 'Verified Teacher'}
   end
 
+  def self.copy_to_script(original_lesson, destination_script)
```

Comment:
Do we want levels to be shared between scripts for the AIML standalone and CSD script? It seems like this leads to many of the things Josh called out for reasons why teachers want to reset progress. Should this only allow you to copy lessons to a different course version within the CSF curriculum umbrella? 

At least it would be good to add a comment at the top of this method explaining some of the details of this method for people who may come on it later

Reply:
> Do we want levels to be shared between scripts for the AIML standalone and CSD script?

No, we do not.

> It seems like this leads to many of the things Josh called out for reasons why teachers want to reset progress.

I don't think this is true, because progress is not shared between scripts. in the past we have had a vague goal of discouraging reuse of levels across scripts, which we might need to revisit to see if it is still a goal. in the mean time, I like the general idea of restricting use of this feature to when its necessary. 

>  Should this only allow you to copy lessons to a different course version within the CSF curriculum umbrella?

During the curriculum meeting on Tuesday, Hannah mentioned using a "staging" script and moving lessons one-by-one into a pilot script as a way of controlling access to the later lessons in the unit. I don't think we should eliminate this use case (which was possible to do using script dsl) until/unless we have a clear plan to support "visible after" or another similar feature for granular control over which lessons are visible.

Reply:
> At least it would be good to add a comment at the top of this method

Will do!

> What do you think about adding a check which ensures (via raising) that if old and new script both have a course version, that the version year matches?

This seems reasonable to me, I'll add this

> Do we want levels to be shared between scripts for the AIML standalone and CSD script? It seems like this leads to many of the things Josh called out for reasons why teachers want to reset progress. Should this only allow you to copy lessons to a different course version within the CSF curriculum umbrella?

I'm not super familiar with the progress details here -- why are the AI/ML scripts different than the CSF courses?

Reply:
some history here: at one point we said we wanted to move in the direction of levels only existing in one script, to simplify various things, including the whole idea that one would have to "find and add" a level into a script rather than just creating it in the place that it needs to be and copying/moving it when necessary. however, Mike objected because this would mean a lot of extra work for CSF where lots of levels are already shared, and generally need to be updated together. IIRC we reached a compromise that we could do this (prevent levels from being shared across scripts) if we built a good bulk-editing tool.

So, I'm not sure AI/ML is all that different, but rather it is not getting grandfathered into the option of sharing levels between scripts like CSF was. I imagine whoever maintains CSD will eventually run into the same headaches of having to update the same level in multiple places, so maybe it is not fair to exclude them unless we are committed to the goal of "one script per level".

Perhaps we can talk at stand up about whether the goal of preventing levels from being shared across scripts is a good one?

Reply:
Sorry, I hadn't refreshed to see Dave's comment! 

Going back to the standalone/CSD AI script, that would be copied with script copy right? It definitely adjusts priority of script copy (or adding level copy to lesson copy) though if they don't want to share levels.

Reply:
> Sorry, I hadn't refreshed to see Dave's comment!

No worries, your question was still relevant!
 
> Going back to the standalone/CSD AI script, that would be copied with script copy right? It definitely adjusts priority of script copy (or adding level copy to lesson copy) though if they don't want to share levels.

Yes, we'd still need script copy (or adding level copy to lesson copy) for hard launch.

Reply:
AIML could be copied into CSD using either script copy. or lesson copy depending on what is ready. But we need to be able to do the copy before hard launch.